### PR TITLE
Add ReadOnlySequence support to Utf8JsonReader with unit tests

### PR DESCRIPTION
**Pull request by ahsonkhan**
**TODO:**
- ~Improve multi-segment test coverage (particularly invalid JSON)~ https://github.com/dotnet/corefx/issues/33584
- ~Refactor duplicate code between single segment and multi-segment (where possible)~ https://github.com/dotnet/corefx/issues/33585

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 1]
```diff
@@ -0,0 +1,1677 @@
...
+                        }
+                        return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, localBuffer[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpaceMultiSegment()
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: This looks identical to the Span one. Am I missing something?

Come to think of it, a lot of these methods don't seem to talk about the sequence object, or call GetNextSpan, so they should be (seemingly) identical to the existing ones.

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Fixed. Removing WIP

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    // Inspired by BitArray
+    // https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/BitArray.cs
+    internal struct CustomUncheckedBitArray
+    {
+        private int[] _array;
+
+        public int Length { get; private set; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public CustomUncheckedBitArray(int bitLength, int integerLength)
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: You're taking both the bitlength and the integerLength to avoid needing to bit shift the bitLength?  The only call site for this constructor uses constants... is the JIT not able to do constant propagation when the ctor gets inlined?

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                }
+                else
+                {
+                    newValue &= ~(1 << extraBits);
+                }
+                _array[elementIndex] = newValue;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: Is this necessary?  In what situation would the JIT otherwise inline this huge method?

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                index++;
+            }
+
+            newints += ((index - 1) >> 5) + 1;   // (value + 31) / 32 without overflow
+
+            if (newints > _array.Length)
+            {
+                var newArray = new int[newints];
+                _array.AsSpan().CopyTo(newArray);
+                _array = newArray;
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: Why can't this just use Array.Resize?

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    // Inspired by BitArray
+    // https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/BitArray.cs
+    internal struct CustomUncheckedBitArray
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: What does this have to do with ReadOnlySequence?

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                // clear high bit values in the last int
+                int last = (Length - 1) >> 5;
+                Div32Rem(Length, out int bits);
+                if (bits > 0)
+                {
+                    _array[last] &= (1 << bits) - 1;
+                }
+
+                // clear remaining int values
+                _array.AsSpan(last + 1, newints - last - 1).Clear();
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: Why is this needed?

File: src/System.Text.Json/src/System/Text/Json/JsonReaderState.cs [hunk 7]
```diff
@@ -79,9 +85,11 @@ public JsonReaderState(int maxDepth = StackFreeMaxDepth, JsonReaderOptions optio
             _previousTokenType = default;
             _readerOptions = options;
 
-            // Only allocate the stack if the user reads a JSON payload beyond the depth that the _stackFreeContainer can handle.
+            // Only allocate if the user reads a JSON payload beyond the depth that the _allocationFreeContainer can handle.
             // This way we avoid allocations in the common, default cases, and allocate lazily.
-            _stack = null;
+            _bitArray = default;
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/JsonReaderState.cs]: Though this also makes JsonState larger, because CustomUncheckedBitArray is larger the reference that's stored today.  And that means any heap object storing JsonState (e.g. an async state machine) is now larger as a result, even if the stack is never used.

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 8]
```diff
@@ -0,0 +1,2408 @@
...
+            {
+                _nextPosition = default;
+                _currentPosition = default;
+                _isLastSegment = isFinalBlock;
+                _isSingleSegment = true;
+            }
+            else
+            {
+                _nextPosition = jsonData.Start;
+                if (_buffer.Length == 0)
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Why is this check needed?  If it's because the first segment of the sequence might be empty, do we not also need to care about a case where the first segment isn't empty but its start position is pointing to the end of the first segment?  Or is that not possible?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 9]
```diff
@@ -0,0 +1,2408 @@
...
+
+        Done:
+            return retVal;
+
+        ReadFirstToken:
+            retVal = ReadFirstTokenMultiSegment(first);
+            goto Done;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Really?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 10]
```diff
@@ -0,0 +1,2408 @@
...
+
+                    leftToMatch = leftToMatch.Slice(span.Length);
+                    alreadyMatched = span.Length;
+                }
+            }
+        Throw:
+            _totalConsumed = prevTotalConsumed;
+            consumed = default;
+            ThrowInvalidLiteralMultiSegment(span);
+            return false;
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: This is the only call site for ThrowInvalidLiteralMultiSegment.  Why not have it return the exception to throw and then use `throw` here rather than having the unreachable `return false`?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 11]
```diff
@@ -0,0 +1,2408 @@
...
+                        indexOfFirstMismatch = i;
+                        break;
+                    }
+                }
+                else
+                {
+                    indexOfFirstMismatch = i;
+                    break;
+                }
+            }
```
Comment (stephentoub) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Rather than branching on every iteration, would it be better to iterate to Min(span.Length, literal.Length), and then if there's no mismatch up until that point and one is longer than the other, that's the position of the first mismatch?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 12]
```diff
@@ -0,0 +1,2408 @@
...
+                ReadOnlySpan<byte> localBuffer = _buffer;
+                for (; _consumed < localBuffer.Length; _consumed++)
+                {
+                    byte val = localBuffer[_consumed];
+
+                    // JSON RFC 8259 section 2 says only these 4 characters count, not all of the Unicode defintions of whitespace.
+                    if (val != JsonConstants.Space &&
+                        val != JsonConstants.CarriageReturn &&
+                        val != JsonConstants.LineFeed &&
+                        val != JsonConstants.Tab)
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Rather than repeat all of this, would

```C#
private void SkipWhiteSpaceMultiSegment()
{
    while (true)
    {
        SkipWhitespace();

        if (_consumed < _buffer.Length)
        {
            break;
        }

        if (!GetNextSpan())
        {
            break;
        }
    }
}
```

work?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 13]
```diff
@@ -0,0 +1,2408 @@
...
+        Throw:
+            _totalConsumed = prevTotalConsumed;
+            consumed = default;
+            ThrowInvalidLiteralMultiSegment(span);
+            return false;
+        }
+
+        private int FindMismatch(ReadOnlySpan<byte> span, ReadOnlySpan<byte> literal)
+        {
+            Debug.Assert(span.Length > 0 && span[0] == literal[0]);
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: I don't think this assert holds.

If a Sequence has a Span break between tr and Ue (case misalignment on the U) then on resumption "Ue" doesn't start with "ue" and "ue" doesn't start with "Ue", so you'll call FindMismatch where 0 is the error position.

So the sequence one needs to still check index 0.  Unless I've missed something.

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 14]
```diff
@@ -0,0 +1,2408 @@
...
+                    }
+
+                    leftToMatch = leftToMatch.Slice(span.Length);
+                    alreadyMatched = span.Length;
+                }
+            }
+        Throw:
+            _totalConsumed = prevTotalConsumed;
+            consumed = default;
+            ThrowInvalidLiteralMultiSegment(span);
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Is this throw correct if the sequence had to move to another span while reading the literal?  If I followed the flow correctly `span` will be pointing at the first letter in the last loaded Span.

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs [hunk 15]
```diff
@@ -0,0 +1,2408 @@
...
+            Debug.Assert(_buffer[_consumed] == JsonConstants.Quote);
+
+            // Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localBuffer = _buffer.Slice(_consumed + 1);
+
+            // Vectorized search for either quote, backslash, or any control character.
+            // If the first found byte is a quote, we have reached an end of string, and
+            // can avoid validation.
+            // Otherwise, in the uncommon case, iterate one character at a time and validate.
+            int idx = localBuffer.IndexOfQuoteOrAnyControlOrBaskSlash();
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: I know it's an existing problem, but I've seen this now and can't unsee it: The method says Bask Slash instead of Back Slash.  And now I'm imagining slashes lying on deck chairs, basking in the sun.  (I'm concerned someone's going to lay on them, since they look like the deck chairs themselves)

Comment (stephentoub) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Lol.  Literally.

File: src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs]: It doesn't. I just happen to add ReadOnlySequeunce on top of the changes from this PR: https://github.com/dotnet/corefx/pull/33378

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 15 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: > And now I'm imagining slashes lying on deck chairs, basking in the sun. (I'm concerned someone's going to lay on them, since they look like the deck chairs themselves)

But the question remains... what's their index?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 9 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Will remove. This method grew larger than what I initially had it as.

**Comment by ahsonkhan** [comment]
Any other feedback?

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 8 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: > Why is this check needed? If it's because the first segment of the sequence might be empty, 

Yes.

> do we not also need to care about a case where the first segment isn't empty but its start position is pointing to the end of the first segment? Or is that not possible?

We don't need to check for that explicitly since the length check covers that case already. We are getting `_buffer` by calling `ReadOnlySequence.First.Span` and First already does the slicing based on start position. So, if we have a segment with the start position at the end, the span we get will have length 0.
https://github.com/dotnet/corefx/blob/ec7b5ed742b0bb5a233d31f472be082ce57e63ad/src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs#L112-L117

cc @pakrym

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 13 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Yep, you are right. This assert only holds for the first call to `FindMismatch` (with the initial span), prior to GetNextSpan.

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 14 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: > will be pointing at the first letter in the last loaded Span.

Yep. Is that a concern? Given this is a literal of known constant size, should I stackalloc and track the entire mismatched literal text we have seen thus far instead?

```text
System.Text.JsonLab.JsonReaderException : Expected a 'true' value, but instead got 'U'. LineNumber: 1 | BytePosition: 3.
```

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: Fixed:
```text
Message: System.Text.JsonLab.JsonReaderException : Expected a 'true' value, but instead got 'trU'. LineNumber: 1 | BytePosition: 2.
```

File: src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs (hunk 11 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs]: This is on the exception code path, so the performance doesn't matter too much here (especially since the literal is at most length 5). But will fix.

**Comment by ahsonkhan** [comment]
If there is no other significant PR feedback, I would like to merge this in so it can be consumed upstream.

**Comment by ahsonkhan** [comment]
Fixes https://github.com/dotnet/corefx/issues/33296
